### ISSUE TITLE
Is NSwag officially dead?

### ISSUE THREAD

**Issue by euroUK** [RicoSuter/NSwag#4884]
Sorry if it sounds a little bit sad, but can we now assume that project is officially dead?

There were breaking changes in v14 so half of generation schemas not working:

typesToOpenApi
webApiToOpenApi

https://github.com/RicoSuter/NSwag/issues/4524

So there are several issues like https://github.com/RicoSuter/NSwag/issues/4731 or https://github.com/RicoSuter/NSwag/issues/4659 where users trying to migrate and get huge problems ike PC hanging since the start of the year.

There is issue https://github.com/RicoSuter/NSwag/issues/4404 from that but there is no responce from March.

In my company we had pipeline with typesToOpenApi because we have shared models in separate solution and right now I don't see any  easy solutions to migrate to .Net 8.

For me now I'm going to write my own generator because there is no even timeline to fix those problems.

**Comment by MoonStorm** [comment]
It looks like it. And the list of open issues keeps on growing. The generator is filled with hundreds of options which amazingly don't solve basic problems (e.g. #3851). We've been running serialization workarounds for years. V13 already had plenty of issues before V14 came along to add even more salt to the wounds. Casing is now all over the place, unwanted or improper validations are still here to stay, it's all a big mess. In my opinion, a reboot of this project, definitely with better stewards, is  definitely needed at this point.

**Comment by SebastianStehle** [comment]
I would say that Nswag is a little bit too complex. 

But I am also tired about the complains in Open Source all the time. Rico has probably generated Millions in value already and there are very little commits from other developers and large Organizations.

**Comment by lahma** [comment]
Latest release is from March so I wouldn't jump to conclusions about project being dead. I think it's lovely that both @euroUK and @MoonStorm are complaining about issues and I fail to see any actual contributions from you. You surely are already getting your money's worth.

**Comment by MoonStorm** [comment]
> Latest release is from March so I wouldn't jump to conclusions about project being dead. I think it's lovely that both @euroUK and @MoonStorm are complaining about issues and I fail to see any actual contributions from you. You surely are already getting your money's worth.

I'm sorry I couldn't find the time to understand and contribute to the project, but there are close to 100 PRs waiting to be triaged and integrated into the code base. Many of them are years old already and the older they get, the more difficult it is to integrate them.

**Comment by lahma** [comment]
Please also consider the burden from the PRs. They are made with best intentions, but they require review (time away from your life) and also they come with a risk. There have been some regressions recently based on improvements that have had ill-effects on other areas - the possible combinations of usage scenarios is mind-blowing.

So when you ask to merge/propose a PR to close an issue the end result might be 10 more issues being opened because something else is now broken. By merging a well-intentioned PR you might spent time trying to fix and release a new version (hopefully) fixing regressions and not breaking anything more.

Of course there are "easy" ones that (at least) seem like no-brainers, but they take effort.

**Comment by MoonStorm** [comment]
That's very true. You do need time to approve/reject/adjust PRs. Many of them might target a very specific case that might not be in line with the path the project is supposed to be on. Reason why for example we have a million different options in the code generation, many of them overlapping on not solving basic issues. 

But here lies the problem as well. You can't ask for contributions if you don't even have the people or the time to get them sorted. A project where you have the volunteers who offer their work but you don't have time to manage them is pretty much a dead project.

**Comment by lahma** [comment]
> That's very true. You do need time to approve/reject/adjust PRs. Many of them might target a very specific case that might not be in line with the path the project is supposed to be on. Reason why for example we have a million different options in the code generation, many of them overlapping on not solving basic issues.
> 
> But here lies the problem as well. You can't ask for contributions if you don't even have the people or the time to get them sorted. A project where you have the volunteers who offer their work but you don't have time to manage them is pretty much a dead project.

Do you have plans on starting a similar project with management that fills your expectations? I think there would be plentiful of people consuming it and creating a backlog for you to work on (please make it free).

**Comment by MoonStorm** [comment]
No. But I hope somebody will soon. Even now I'm starring at a freshly new generated set of DTOs, where several of them were generated with a small case name, in C#, with the compiler obviously blowing up on such a simple coding style violation. 

Lack of proper tests, lack of a solid foundation, where the quantity of features prevailed over quality. A project that was caught off guard by its popularity. Such a shame.

**Comment by lahma** [comment]
> No. But I hope somebody will soon. Even now I'm staring at a freshly new generated set of DTOs, where several of them were generated with a small case name, in C#, with the compiler obviously blowing up on such a simple coding style violation. 
> 
> Lack of proper tests, lack of a solid foundation, where the quantity of features prevailed over quality. A project that was caught off guard by its popularity. Such a shame.

Now I believe the audience would love to see how proper software should be built. Can you share some of your work?

**Comment by MoonStorm** [comment]
The audience is busy logging issues.

**Comment by MoonStorm** [comment]
Ummm, I think we all know by now how "proper software" should be built. Or don't we? A quick google search: [SOLID Principles](https://www.geeksforgeeks.org/solid-principle-in-programming-understand-with-real-life-examples/), [Design Patterns](https://sourcemaking.com/design_patterns), [TDD](https://martinfowler.com/bliki/TestDrivenDevelopment.html), [Coding Styles](https://www.geeksforgeeks.org/coding-standards-and-guidelines/), etc.

**Comment by MoonStorm** [comment]
I don't even see an editorconfig file in this project with the code rules properly defined!

**Comment by MoonStorm** [comment]
Oh I see, now you're.... "minimizing" my comments 🤣
Isn't it sad when even such a small amount of power get to people's head? 
Yup, these are definitely the right people in charge of this project. Good luck over there!

**Comment by lahma** [comment]
> Oh I see, now you're.... "minimizing" my comments 🤣 Isn't it sad when even such a small amount of power get to people's head? Yup, these are definitely the right people in charge of this project. Good luck over there!

Just because your condescending attitude is not welcome here.

**Comment by MoonStorm** [comment]
Ah don't worry about that, don't really wanna be here. I wish this is gonna be restarted by more talented people who do have more time to care for it. The idea behind this project is brilliant, the implementation is simply sub-par for its popularity, as I said, reason for the neverending list of unsolved issues that keep on piling up.

**Comment by lahma** [comment]
Thank you for your input. Like you've discovered, it's hard to find the needed talented people. Me and you wouldn't clearly make the cut.

**Comment by euroUK** [comment]
@lahma 
Why so toxic?

I just wanted to decide to remove NSwag from toolchain completly or wait a little bit and may be things will be fixed.

The reasons I'm not contributing to Open Source are obvious - I'm working full time as software developer and I don't have enough free time to properly manage open source solutions. My code is enough for our company needs but you need to invest 3x more time to properly release it for everyone. I don't blame if Rico has no time, It would not take too much time to generate typesToOpenApi for our toolchain. I just don't want to reinvent the weel.

Moreover, the open source initiative appears to be dead because the only alive solutions are those supported or developed by large corporations.

**Comment by jamesrwaugh** [comment]
I think yes, unfortunately. NSwag was good but it's reached it's internal complexity limit and the open-to-closed issue count is not promising.

Nowadays Microsoft's Koita is an option.

https://github.com/microsoft/kiota

**Comment by PooSham** [comment]
New commits were made 2 weeks ago and version bumped to 14.0.8. Can we close this issue now or is there still a reason to believe the project is basically dead?

**Comment by eolamisan** [comment]
I would say it's still a problem because of the **1.8k open issues**.
4 commits per month not going to solve the problem. (Even if each commit were to clear 10 issues.)
It looks like with current velocity it would require years of effort just to go through the current backlog.

It seems to be the sad reality of many a popular OSS project lately, without financial backing.
Also, it's not fair to expect Rico to work through 1.8 issues (even paid, let alone free).
I would agree, Rico has generated millions in value, but here we are. 

Code frequency
![image](https://github.com/RicoSuter/NSwag/assets/4191912/97b9bd03-73ed-431a-805c-26b4681b107c)

https://github.com/RicoSuter/NSwag/graphs/commit-activity

**Comment by awdorrin** [comment]
With only a few months before .Net 6 goes end-of-life, I have to wonder if it is time to move on from NSwag.
I've spent several hours today, trying to understand why NSwag is unrolling DTO classes that were not being unrolled under .net 6, and have gotten nowhere.
An issue I created back in February (https://github.com/RicoSuter/NSwag/issues/4791) has no responses, other than my own and one other asking if I found a work-around.
I know we can't be the only ones encountering this with Typescript....

I have experimented a bit with OpenAPITools/openapi-generator-cli, in a smaller project, and while it isn't perfect, it doesn't do this class unrolling from REST APIs that is our biggest issue with NSwag.

**Comment by url00** [comment]
> With only a few months before .Net 6 goes end-of-life, I have to wonder if it is time to move on from NSwag. I've spent several hours today, trying to understand why NSwag is unrolling DTO classes that were not being unrolled under .net 6, and have gotten nowhere. An issue I created back in February (#4791) has no responses, other than my own and one other asking if I found a work-around. I know we can't be the only ones encountering this with Typescript....
> 
> I have experimented a bit with OpenAPITools/openapi-generator-cli, in a smaller project, and while it isn't perfect, it doesn't do this class unrolling from REST APIs that is our biggest issue with NSwag.

We also made the switch to OpenAPITools/openapi-generator-cli, keeping the old working generated NSwag API in its current state while we transition over. It's too bad but at this point I think it's the best option. So far everything works as expected.

**Comment by vanhoutenbos** [comment]
NSwag perse doesn't seem dead to me at all, just the support for specifically console based tool (https://www.nuget.org/packages/NSwag.ConsoleCore)

Anything above .NET 6 seems to be really wonky

**Comment by lahma** [comment]
Closing this as it's more productive to address more specific issues, hopefully with pull requests.
